c10/util/intrusive_ptr.h

```diff
@@ -207,15 +207,28 @@ class intrusive_ptr final {
     target_ = NullType::singleton();
   }
 
-  // This constructor will not increase the ref counter for you.
-  // This is not public because we shouldn't make intrusive_ptr out of raw
-  // pointers except from inside the make_intrusive() and
-  // weak_intrusive_ptr::lock() implementations
-  explicit intrusive_ptr(TTarget* target) noexcept : target_(target) {}
-
  public:
   using element_type = TTarget;
 
+  // This constructor will by_default increase the ref counter for you.
+  // This is public because:
+  // 1. Boost intrusive_ptr has the same public API
+  //    https://www.boost.org/doc/libs/1_74_0/boost/smart_ptr/intrusive_ptr.hpp
+  // 2. Pybind11 custom smart pointer requires pointer/raw pointer constructor
+  //    to be a part of public API
+  //    https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#custom-smart-pointers
+  explicit intrusive_ptr(TTarget* target, bool add_ref=true) noexcept : target_(target) {
```

Comment:
I would rather not make this public API. Using the regular intrusive_ptr APIs should be safe and you shouldn't be able to shoot yourself in the foot with them. We have the `raw::incref()` APIs for people who know what they're doing that opt in to dangerous behavior by using the `raw` namespace.

Also, defaulting this to true disables the protection we have against making intrusive_ptrs to stack objects. It is intentional that the intrusive_ptr does not increment refcount but the `make_intrusive` does it. See the comment about "Stack allocated intrusive_ptr_target safety" at the top of the file.